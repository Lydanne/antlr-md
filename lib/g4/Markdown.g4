grammar Markdown;

markdown: block* EOF?;

block: inlineHeader | inlineText | blockCode | NL;

inlineHeader: header textContent NL?;

header: HEADER;

inlineText: textContent NL?;

textContent: (inlineCode | str)+;

blockCode: BLOCK_CODE NL?;

str: CHAR+;

inlineCode: INLINE_CODE;

HEADER: HASH+ WS;

CHAR: ~[\r\n];

INLINE_CODE: BACKTICK ~[\r\n`]+ BACKTICK;

BLOCK_CODE: '```' ~[`]+ '```';

NL: '\r'? '\n';

fragment BACKTICK: '`';

fragment WS: [ \t];

fragment HASH: '#';