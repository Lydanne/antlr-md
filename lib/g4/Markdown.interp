token literal names:
null
null
null
null
null
null

token symbolic names:
null
HEADER
CHAR
INLINE_CODE
BLOCK_CODE
NL

rule names:
markdown
block
inlineHeader
inlineText
textContent
blockCode
text
inlineCode


atn:
[4, 1, 5, 61, 2, 0, 7, 0, 2, 1, 7, 1, 2, 2, 7, 2, 2, 3, 7, 3, 2, 4, 7, 4, 2, 5, 7, 5, 2, 6, 7, 6, 2, 7, 7, 7, 1, 0, 5, 0, 18, 8, 0, 10, 0, 12, 0, 21, 9, 0, 1, 0, 3, 0, 24, 8, 0, 1, 1, 1, 1, 1, 1, 1, 1, 3, 1, 30, 8, 1, 1, 2, 3, 2, 33, 8, 2, 1, 2, 1, 2, 1, 2, 1, 3, 3, 3, 39, 8, 3, 1, 3, 1, 3, 1, 4, 1, 4, 4, 4, 45, 8, 4, 11, 4, 12, 4, 46, 1, 5, 3, 5, 50, 8, 5, 1, 5, 1, 5, 1, 6, 4, 6, 55, 8, 6, 11, 6, 12, 6, 56, 1, 7, 1, 7, 1, 7, 0, 0, 8, 0, 2, 4, 6, 8, 10, 12, 14, 0, 0, 63, 0, 19, 1, 0, 0, 0, 2, 29, 1, 0, 0, 0, 4, 32, 1, 0, 0, 0, 6, 38, 1, 0, 0, 0, 8, 44, 1, 0, 0, 0, 10, 49, 1, 0, 0, 0, 12, 54, 1, 0, 0, 0, 14, 58, 1, 0, 0, 0, 16, 18, 3, 2, 1, 0, 17, 16, 1, 0, 0, 0, 18, 21, 1, 0, 0, 0, 19, 17, 1, 0, 0, 0, 19, 20, 1, 0, 0, 0, 20, 23, 1, 0, 0, 0, 21, 19, 1, 0, 0, 0, 22, 24, 5, 0, 0, 1, 23, 22, 1, 0, 0, 0, 23, 24, 1, 0, 0, 0, 24, 1, 1, 0, 0, 0, 25, 30, 3, 4, 2, 0, 26, 30, 3, 6, 3, 0, 27, 30, 3, 10, 5, 0, 28, 30, 5, 5, 0, 0, 29, 25, 1, 0, 0, 0, 29, 26, 1, 0, 0, 0, 29, 27, 1, 0, 0, 0, 29, 28, 1, 0, 0, 0, 30, 3, 1, 0, 0, 0, 31, 33, 5, 5, 0, 0, 32, 31, 1, 0, 0, 0, 32, 33, 1, 0, 0, 0, 33, 34, 1, 0, 0, 0, 34, 35, 5, 1, 0, 0, 35, 36, 3, 8, 4, 0, 36, 5, 1, 0, 0, 0, 37, 39, 5, 5, 0, 0, 38, 37, 1, 0, 0, 0, 38, 39, 1, 0, 0, 0, 39, 40, 1, 0, 0, 0, 40, 41, 3, 8, 4, 0, 41, 7, 1, 0, 0, 0, 42, 45, 3, 14, 7, 0, 43, 45, 3, 12, 6, 0, 44, 42, 1, 0, 0, 0, 44, 43, 1, 0, 0, 0, 45, 46, 1, 0, 0, 0, 46, 44, 1, 0, 0, 0, 46, 47, 1, 0, 0, 0, 47, 9, 1, 0, 0, 0, 48, 50, 5, 5, 0, 0, 49, 48, 1, 0, 0, 0, 49, 50, 1, 0, 0, 0, 50, 51, 1, 0, 0, 0, 51, 52, 5, 4, 0, 0, 52, 11, 1, 0, 0, 0, 53, 55, 5, 2, 0, 0, 54, 53, 1, 0, 0, 0, 55, 56, 1, 0, 0, 0, 56, 54, 1, 0, 0, 0, 56, 57, 1, 0, 0, 0, 57, 13, 1, 0, 0, 0, 58, 59, 5, 3, 0, 0, 59, 15, 1, 0, 0, 0, 9, 19, 23, 29, 32, 38, 44, 46, 49, 56]