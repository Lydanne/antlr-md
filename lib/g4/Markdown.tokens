HEADER=1
CHAR=2
INLINE_CODE=3
BLOCK_CODE=4
NL=5
