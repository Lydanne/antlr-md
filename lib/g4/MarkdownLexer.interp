token literal names:
null
null
null
null
null
null
'#'
null
'*'
null
'.'
null
'_'
'`'
'\\*'
'**'
'__'
'```'
null

token symbolic names:
null
BOLD_TEXT
ITALIC_TEXT
INLINE_CODE
CODE_BLOCK
TEXT
HASH
HASH_WS
STAR
DIGITS
DOT
WS
UNDERSCORE
BACKTICK
ASTERISK
DOUBLE_ASTERISK
DOUBLE_UNDERSCORE
TRIPLE_BACKTICK
NL

rule names:
BOLD_TEXT
ITALIC_TEXT
INLINE_CODE
CODE_BLOCK
TEXT
HASH
HASH_WS
STAR
DIGITS
DOT
WS
UNDERSCORE
BACKTICK
ASTERISK
DOUBLE_ASTERISK
DOUBLE_UNDERSCORE
TRIPLE_BACKTICK
NL

channel names:
DEFAULT_TOKEN_CHANNEL
HIDDEN

mode names:
DEFAULT_MODE

atn:
[3, 51485, 51898, 1421, 44986, 20307, 1543, 60043, 49729, 2, 20, 141, 8, 1, 4, 2, 9, 2, 4, 3, 9, 3, 4, 4, 9, 4, 4, 5, 9, 5, 4, 6, 9, 6, 4, 7, 9, 7, 4, 8, 9, 8, 4, 9, 9, 9, 4, 10, 9, 10, 4, 11, 9, 11, 4, 12, 9, 12, 4, 13, 9, 13, 4, 14, 9, 14, 4, 15, 9, 15, 4, 16, 9, 16, 4, 17, 9, 17, 4, 18, 9, 18, 4, 19, 9, 19, 3, 2, 3, 2, 6, 2, 42, 10, 2, 13, 2, 14, 2, 43, 3, 2, 3, 2, 3, 2, 3, 2, 6, 2, 50, 10, 2, 13, 2, 14, 2, 51, 3, 2, 3, 2, 5, 2, 56, 10, 2, 3, 3, 3, 3, 6, 3, 60, 10, 3, 13, 3, 14, 3, 61, 3, 3, 3, 3, 3, 3, 3, 3, 6, 3, 68, 10, 3, 13, 3, 14, 3, 69, 3, 3, 3, 3, 5, 3, 74, 10, 3, 3, 4, 3, 4, 6, 4, 78, 10, 4, 13, 4, 14, 4, 79, 3, 4, 3, 4, 3, 5, 3, 5, 7, 5, 86, 10, 5, 12, 5, 14, 5, 89, 11, 5, 3, 5, 3, 5, 3, 5, 3, 5, 3, 6, 6, 6, 96, 10, 6, 13, 6, 14, 6, 97, 3, 7, 3, 7, 3, 8, 6, 8, 103, 10, 8, 13, 8, 14, 8, 104, 3, 8, 3, 8, 3, 9, 3, 9, 3, 10, 6, 10, 112, 10, 10, 13, 10, 14, 10, 113, 3, 11, 3, 11, 3, 12, 3, 12, 3, 13, 3, 13, 3, 14, 3, 14, 3, 15, 3, 15, 3, 15, 3, 16, 3, 16, 3, 16, 3, 17, 3, 17, 3, 17, 3, 18, 3, 18, 3, 18, 3, 18, 3, 19, 5, 19, 138, 10, 19, 3, 19, 3, 19, 3, 87, 2, 2, 20, 3, 2, 3, 5, 2, 4, 7, 2, 5, 9, 2, 6, 11, 2, 7, 13, 2, 8, 15, 2, 9, 17, 2, 10, 19, 2, 11, 21, 2, 12, 23, 2, 13, 25, 2, 14, 27, 2, 15, 29, 2, 16, 31, 2, 17, 33, 2, 18, 35, 2, 19, 37, 2, 20, 3, 2, 8, 5, 2, 12, 12, 15, 15, 44, 44, 5, 2, 12, 12, 15, 15, 97, 97, 5, 2, 12, 12, 15, 15, 98, 98, 7, 2, 12, 12, 15, 15, 37, 37, 44, 44, 97, 98, 3, 2, 50, 59, 4, 2, 11, 11, 34, 34, 2, 152, 2, 3, 3, 2, 2, 2, 2, 5, 3, 2, 2, 2, 2, 7, 3, 2, 2, 2, 2, 9, 3, 2, 2, 2, 2, 11, 3, 2, 2, 2, 2, 13, 3, 2, 2, 2, 2, 15, 3, 2, 2, 2, 2, 17, 3, 2, 2, 2, 2, 19, 3, 2, 2, 2, 2, 21, 3, 2, 2, 2, 2, 23, 3, 2, 2, 2, 2, 25, 3, 2, 2, 2, 2, 27, 3, 2, 2, 2, 2, 29, 3, 2, 2, 2, 2, 31, 3, 2, 2, 2, 2, 33, 3, 2, 2, 2, 2, 35, 3, 2, 2, 2, 2, 37, 3, 2, 2, 2, 3, 55, 3, 2, 2, 2, 5, 73, 3, 2, 2, 2, 7, 75, 3, 2, 2, 2, 9, 83, 3, 2, 2, 2, 11, 95, 3, 2, 2, 2, 13, 99, 3, 2, 2, 2, 15, 102, 3, 2, 2, 2, 17, 108, 3, 2, 2, 2, 19, 111, 3, 2, 2, 2, 21, 115, 3, 2, 2, 2, 23, 117, 3, 2, 2, 2, 25, 119, 3, 2, 2, 2, 27, 121, 3, 2, 2, 2, 29, 123, 3, 2, 2, 2, 31, 126, 3, 2, 2, 2, 33, 129, 3, 2, 2, 2, 35, 132, 3, 2, 2, 2, 37, 137, 3, 2, 2, 2, 39, 41, 5, 31, 16, 2, 40, 42, 10, 2, 2, 2, 41, 40, 3, 2, 2, 2, 42, 43, 3, 2, 2, 2, 43, 41, 3, 2, 2, 2, 43, 44, 3, 2, 2, 2, 44, 45, 3, 2, 2, 2, 45, 46, 5, 31, 16, 2, 46, 56, 3, 2, 2, 2, 47, 49, 5, 33, 17, 2, 48, 50, 10, 3, 2, 2, 49, 48, 3, 2, 2, 2, 50, 51, 3, 2, 2, 2, 51, 49, 3, 2, 2, 2, 51, 52, 3, 2, 2, 2, 52, 53, 3, 2, 2, 2, 53, 54, 5, 33, 17, 2, 54, 56, 3, 2, 2, 2, 55, 39, 3, 2, 2, 2, 55, 47, 3, 2, 2, 2, 56, 4, 3, 2, 2, 2, 57, 59, 5, 29, 15, 2, 58, 60, 10, 2, 2, 2, 59, 58, 3, 2, 2, 2, 60, 61, 3, 2, 2, 2, 61, 59, 3, 2, 2, 2, 61, 62, 3, 2, 2, 2, 62, 63, 3, 2, 2, 2, 63, 64, 5, 29, 15, 2, 64, 74, 3, 2, 2, 2, 65, 67, 5, 25, 13, 2, 66, 68, 10, 3, 2, 2, 67, 66, 3, 2, 2, 2, 68, 69, 3, 2, 2, 2, 69, 67, 3, 2, 2, 2, 69, 70, 3, 2, 2, 2, 70, 71, 3, 2, 2, 2, 71, 72, 5, 25, 13, 2, 72, 74, 3, 2, 2, 2, 73, 57, 3, 2, 2, 2, 73, 65, 3, 2, 2, 2, 74, 6, 3, 2, 2, 2, 75, 77, 5, 27, 14, 2, 76, 78, 10, 4, 2, 2, 77, 76, 3, 2, 2, 2, 78, 79, 3, 2, 2, 2, 79, 77, 3, 2, 2, 2, 79, 80, 3, 2, 2, 2, 80, 81, 3, 2, 2, 2, 81, 82, 5, 27, 14, 2, 82, 8, 3, 2, 2, 2, 83, 87, 5, 35, 18, 2, 84, 86, 11, 2, 2, 2, 85, 84, 3, 2, 2, 2, 86, 89, 3, 2, 2, 2, 87, 88, 3, 2, 2, 2, 87, 85, 3, 2, 2, 2, 88, 90, 3, 2, 2, 2, 89, 87, 3, 2, 2, 2, 90, 91, 5, 35, 18, 2, 91, 92, 3, 2, 2, 2, 92, 93, 8, 5, 2, 2, 93, 10, 3, 2, 2, 2, 94, 96, 10, 5, 2, 2, 95, 94, 3, 2, 2, 2, 96, 97, 3, 2, 2, 2, 97, 95, 3, 2, 2, 2, 97, 98, 3, 2, 2, 2, 98, 12, 3, 2, 2, 2, 99, 100, 7, 37, 2, 2, 100, 14, 3, 2, 2, 2, 101, 103, 5, 13, 7, 2, 102, 101, 3, 2, 2, 2, 103, 104, 3, 2, 2, 2, 104, 102, 3, 2, 2, 2, 104, 105, 3, 2, 2, 2, 105, 106, 3, 2, 2, 2, 106, 107, 5, 23, 12, 2, 107, 16, 3, 2, 2, 2, 108, 109, 7, 44, 2, 2, 109, 18, 3, 2, 2, 2, 110, 112, 9, 6, 2, 2, 111, 110, 3, 2, 2, 2, 112, 113, 3, 2, 2, 2, 113, 111, 3, 2, 2, 2, 113, 114, 3, 2, 2, 2, 114, 20, 3, 2, 2, 2, 115, 116, 7, 48, 2, 2, 116, 22, 3, 2, 2, 2, 117, 118, 9, 7, 2, 2, 118, 24, 3, 2, 2, 2, 119, 120, 7, 97, 2, 2, 120, 26, 3, 2, 2, 2, 121, 122, 7, 98, 2, 2, 122, 28, 3, 2, 2, 2, 123, 124, 7, 94, 2, 2, 124, 125, 7, 44, 2, 2, 125, 30, 3, 2, 2, 2, 126, 127, 7, 44, 2, 2, 127, 128, 7, 44, 2, 2, 128, 32, 3, 2, 2, 2, 129, 130, 7, 97, 2, 2, 130, 131, 7, 97, 2, 2, 131, 34, 3, 2, 2, 2, 132, 133, 7, 98, 2, 2, 133, 134, 7, 98, 2, 2, 134, 135, 7, 98, 2, 2, 135, 36, 3, 2, 2, 2, 136, 138, 7, 15, 2, 2, 137, 136, 3, 2, 2, 2, 137, 138, 3, 2, 2, 2, 138, 139, 3, 2, 2, 2, 139, 140, 7, 12, 2, 2, 140, 38, 3, 2, 2, 2, 15, 2, 43, 51, 55, 61, 69, 73, 79, 87, 97, 104, 113, 137, 3, 4, 2, 2]