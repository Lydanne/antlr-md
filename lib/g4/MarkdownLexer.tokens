HEADER=1
CHAR=2
INLINE_CODE=3
BLOCK_CODE=4
BACKTICK=5
NL=6
'`'=5
