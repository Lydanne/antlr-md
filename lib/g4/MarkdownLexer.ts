// @ts-nocheck
// Generated from ./lib/g4/Markdown.g4 by ANTLR 4.12.0
// noinspection ES6UnusedImports,JSUnusedGlobalSymbols,JSUnusedLocalSymbols
import {
	ATN,
	ATNDeserializer,
	CharStream,
	DecisionState, DFA,
	Lexer,
	LexerATNSimulator,
	RuleContext,
	PredictionContextCache,
	Token
} from "antlr4";
export default class MarkdownLexer extends Lexer {
	public static readonly HEADER = 1;
	public static readonly CHAR = 2;
	public static readonly INLINE_CODE = 3;
	public static readonly BLOCK_CODE = 4;
	public static readonly NL = 5;
	public static readonly EOF = Token.EOF;

	public static readonly channelNames: string[] = [ "DEFAULT_TOKEN_CHANNEL", "HIDDEN" ];
	public static readonly literalNames: string[] = [  ];
	public static readonly symbolicNames: string[] = [ null, "HEADER", "CHAR", 
                                                    "INLINE_CODE", "BLOCK_CODE", 
                                                    "NL" ];
	public static readonly modeNames: string[] = [ "DEFAULT_MODE", ];

	public static readonly ruleNames: string[] = [
		"HEADER", "CHAR", "INLINE_CODE", "BLOCK_CODE", "NL", "CODE_LANG", "BLOCK_CODE_FLAG", 
		"BACKTICK", "WS", "HASH",
	];


	constructor(input: CharStream) {
		super(input);
		this._interp = new LexerATNSimulator(this, MarkdownLexer._ATN, MarkdownLexer.DecisionsToDFA, new PredictionContextCache());
	}

	public get grammarFileName(): string { return "Markdown.g4"; }

	public get literalNames(): (string | null)[] { return MarkdownLexer.literalNames; }
	public get symbolicNames(): (string | null)[] { return MarkdownLexer.symbolicNames; }
	public get ruleNames(): string[] { return MarkdownLexer.ruleNames; }

	public get serializedATN(): number[] { return MarkdownLexer._serializedATN; }

	public get channelNames(): string[] { return MarkdownLexer.channelNames; }

	public get modeNames(): string[] { return MarkdownLexer.modeNames; }

	public static readonly _serializedATN: number[] = [4,0,5,74,6,-1,2,0,7,
	0,2,1,7,1,2,2,7,2,2,3,7,3,2,4,7,4,2,5,7,5,2,6,7,6,2,7,7,7,2,8,7,8,2,9,7,
	9,1,0,4,0,23,8,0,11,0,12,0,24,1,0,1,0,1,1,1,1,1,2,1,2,4,2,33,8,2,11,2,12,
	2,34,1,2,1,2,1,3,1,3,3,3,41,8,3,1,3,1,3,5,3,45,8,3,10,3,12,3,48,9,3,1,3,
	3,3,51,8,3,1,3,1,3,1,4,3,4,56,8,4,1,4,1,4,1,5,4,5,61,8,5,11,5,12,5,62,1,
	6,1,6,1,6,1,6,1,7,1,7,1,8,1,8,1,9,1,9,1,46,0,10,1,1,3,2,5,3,7,4,9,5,11,
	0,13,0,15,0,17,0,19,0,1,0,4,2,0,10,10,13,13,3,0,10,10,13,13,96,96,3,0,9,
	10,13,13,96,96,2,0,9,9,32,32,75,0,1,1,0,0,0,0,3,1,0,0,0,0,5,1,0,0,0,0,7,
	1,0,0,0,0,9,1,0,0,0,1,22,1,0,0,0,3,28,1,0,0,0,5,30,1,0,0,0,7,38,1,0,0,0,
	9,55,1,0,0,0,11,60,1,0,0,0,13,64,1,0,0,0,15,68,1,0,0,0,17,70,1,0,0,0,19,
	72,1,0,0,0,21,23,3,19,9,0,22,21,1,0,0,0,23,24,1,0,0,0,24,22,1,0,0,0,24,
	25,1,0,0,0,25,26,1,0,0,0,26,27,3,17,8,0,27,2,1,0,0,0,28,29,8,0,0,0,29,4,
	1,0,0,0,30,32,3,15,7,0,31,33,8,1,0,0,32,31,1,0,0,0,33,34,1,0,0,0,34,32,
	1,0,0,0,34,35,1,0,0,0,35,36,1,0,0,0,36,37,3,15,7,0,37,6,1,0,0,0,38,40,3,
	13,6,0,39,41,3,11,5,0,40,39,1,0,0,0,40,41,1,0,0,0,41,42,1,0,0,0,42,46,3,
	9,4,0,43,45,9,0,0,0,44,43,1,0,0,0,45,48,1,0,0,0,46,47,1,0,0,0,46,44,1,0,
	0,0,47,50,1,0,0,0,48,46,1,0,0,0,49,51,3,9,4,0,50,49,1,0,0,0,50,51,1,0,0,
	0,51,52,1,0,0,0,52,53,3,13,6,0,53,8,1,0,0,0,54,56,5,13,0,0,55,54,1,0,0,
	0,55,56,1,0,0,0,56,57,1,0,0,0,57,58,5,10,0,0,58,10,1,0,0,0,59,61,8,2,0,
	0,60,59,1,0,0,0,61,62,1,0,0,0,62,60,1,0,0,0,62,63,1,0,0,0,63,12,1,0,0,0,
	64,65,5,96,0,0,65,66,5,96,0,0,66,67,5,96,0,0,67,14,1,0,0,0,68,69,5,96,0,
	0,69,16,1,0,0,0,70,71,7,3,0,0,71,18,1,0,0,0,72,73,5,35,0,0,73,20,1,0,0,
	0,8,0,24,34,40,46,50,55,62,0];

	private static __ATN: ATN;
	public static get _ATN(): ATN {
		if (!MarkdownLexer.__ATN) {
			MarkdownLexer.__ATN = new ATNDeserializer().deserialize(MarkdownLexer._serializedATN);
		}

		return MarkdownLexer.__ATN;
	}


	static DecisionsToDFA = MarkdownLexer._ATN.decisionToState.map( (ds: DecisionState, index: number) => new DFA(ds, index) );
}