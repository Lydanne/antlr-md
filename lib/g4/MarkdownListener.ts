// @ts-nocheck
// Generated from ./lib/g4/Markdown.g4 by ANTLR 4.12.0

import {ParseTreeListener} from "antlr4";


import { MarkdownContext } from "./MarkdownParser";
import { BlockContext } from "./MarkdownParser";
import { InlineHeaderContext } from "./MarkdownParser";
import { HeaderContext } from "./MarkdownParser";
import { InlineTextContext } from "./MarkdownParser";
import { TextContentContext } from "./MarkdownParser";
import { BlockCodeContext } from "./MarkdownParser";
import { StrContext } from "./MarkdownParser";
import { InlineCodeContext } from "./MarkdownParser";


/**
 * This interface defines a complete listener for a parse tree produced by
 * `MarkdownParser`.
 */
export default class MarkdownListener extends ParseTreeListener {
	/**
	 * Enter a parse tree produced by `MarkdownParser.markdown`.
	 * @param ctx the parse tree
	 */
	enterMarkdown?: (ctx: MarkdownContext) => void;
	/**
	 * Exit a parse tree produced by `MarkdownParser.markdown`.
	 * @param ctx the parse tree
	 */
	exitMarkdown?: (ctx: MarkdownContext) => void;
	/**
	 * Enter a parse tree produced by `MarkdownParser.block`.
	 * @param ctx the parse tree
	 */
	enterBlock?: (ctx: BlockContext) => void;
	/**
	 * Exit a parse tree produced by `MarkdownParser.block`.
	 * @param ctx the parse tree
	 */
	exitBlock?: (ctx: BlockContext) => void;
	/**
	 * Enter a parse tree produced by `MarkdownParser.inlineHeader`.
	 * @param ctx the parse tree
	 */
	enterInlineHeader?: (ctx: InlineHeaderContext) => void;
	/**
	 * Exit a parse tree produced by `MarkdownParser.inlineHeader`.
	 * @param ctx the parse tree
	 */
	exitInlineHeader?: (ctx: InlineHeaderContext) => void;
	/**
	 * Enter a parse tree produced by `MarkdownParser.header`.
	 * @param ctx the parse tree
	 */
	enterHeader?: (ctx: HeaderContext) => void;
	/**
	 * Exit a parse tree produced by `MarkdownParser.header`.
	 * @param ctx the parse tree
	 */
	exitHeader?: (ctx: HeaderContext) => void;
	/**
	 * Enter a parse tree produced by `MarkdownParser.inlineText`.
	 * @param ctx the parse tree
	 */
	enterInlineText?: (ctx: InlineTextContext) => void;
	/**
	 * Exit a parse tree produced by `MarkdownParser.inlineText`.
	 * @param ctx the parse tree
	 */
	exitInlineText?: (ctx: InlineTextContext) => void;
	/**
	 * Enter a parse tree produced by `MarkdownParser.textContent`.
	 * @param ctx the parse tree
	 */
	enterTextContent?: (ctx: TextContentContext) => void;
	/**
	 * Exit a parse tree produced by `MarkdownParser.textContent`.
	 * @param ctx the parse tree
	 */
	exitTextContent?: (ctx: TextContentContext) => void;
	/**
	 * Enter a parse tree produced by `MarkdownParser.blockCode`.
	 * @param ctx the parse tree
	 */
	enterBlockCode?: (ctx: BlockCodeContext) => void;
	/**
	 * Exit a parse tree produced by `MarkdownParser.blockCode`.
	 * @param ctx the parse tree
	 */
	exitBlockCode?: (ctx: BlockCodeContext) => void;
	/**
	 * Enter a parse tree produced by `MarkdownParser.str`.
	 * @param ctx the parse tree
	 */
	enterStr?: (ctx: StrContext) => void;
	/**
	 * Exit a parse tree produced by `MarkdownParser.str`.
	 * @param ctx the parse tree
	 */
	exitStr?: (ctx: StrContext) => void;
	/**
	 * Enter a parse tree produced by `MarkdownParser.inlineCode`.
	 * @param ctx the parse tree
	 */
	enterInlineCode?: (ctx: InlineCodeContext) => void;
	/**
	 * Exit a parse tree produced by `MarkdownParser.inlineCode`.
	 * @param ctx the parse tree
	 */
	exitInlineCode?: (ctx: InlineCodeContext) => void;
}

