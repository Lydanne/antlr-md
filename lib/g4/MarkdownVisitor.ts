// Generated from ./lib/g4/Markdown.g4 by ANTLR 4.9.0-SNAPSHOT


import { ParseTreeVisitor } from "antlr4ts/tree/ParseTreeVisitor";

import { MarkdownContext } from "./MarkdownParser";
import { BlockContext } from "./MarkdownParser";
import { HeaderContext } from "./MarkdownParser";
import { HeaderFlagContext } from "./MarkdownParser";
import { UnorderedListContext } from "./MarkdownParser";
import { OrderedListContext } from "./MarkdownParser";
import { TextLineContext } from "./MarkdownParser";
import { InlineTextContext } from "./MarkdownParser";
import { BlankLineContext } from "./MarkdownParser";


/**
 * This interface defines a complete generic visitor for a parse tree produced
 * by `MarkdownParser`.
 *
 * @param <Result> The return type of the visit operation. Use `void` for
 * operations with no return type.
 */
export interface MarkdownVisitor<Result> extends ParseTreeVisitor<Result> {
	/**
	 * Visit a parse tree produced by `MarkdownParser.markdown`.
	 * @param ctx the parse tree
	 * @return the visitor result
	 */
	visitMarkdown?: (ctx: MarkdownContext) => Result;

	/**
	 * Visit a parse tree produced by `MarkdownParser.block`.
	 * @param ctx the parse tree
	 * @return the visitor result
	 */
	visitBlock?: (ctx: BlockContext) => Result;

	/**
	 * Visit a parse tree produced by `MarkdownParser.header`.
	 * @param ctx the parse tree
	 * @return the visitor result
	 */
	visitHeader?: (ctx: HeaderContext) => Result;

	/**
	 * Visit a parse tree produced by `MarkdownParser.headerFlag`.
	 * @param ctx the parse tree
	 * @return the visitor result
	 */
	visitHeaderFlag?: (ctx: HeaderFlagContext) => Result;

	/**
	 * Visit a parse tree produced by `MarkdownParser.unorderedList`.
	 * @param ctx the parse tree
	 * @return the visitor result
	 */
	visitUnorderedList?: (ctx: UnorderedListContext) => Result;

	/**
	 * Visit a parse tree produced by `MarkdownParser.orderedList`.
	 * @param ctx the parse tree
	 * @return the visitor result
	 */
	visitOrderedList?: (ctx: OrderedListContext) => Result;

	/**
	 * Visit a parse tree produced by `MarkdownParser.textLine`.
	 * @param ctx the parse tree
	 * @return the visitor result
	 */
	visitTextLine?: (ctx: TextLineContext) => Result;

	/**
	 * Visit a parse tree produced by `MarkdownParser.inlineText`.
	 * @param ctx the parse tree
	 * @return the visitor result
	 */
	visitInlineText?: (ctx: InlineTextContext) => Result;

	/**
	 * Visit a parse tree produced by `MarkdownParser.blankLine`.
	 * @param ctx the parse tree
	 * @return the visitor result
	 */
	visitBlankLine?: (ctx: BlankLineContext) => Result;
}

