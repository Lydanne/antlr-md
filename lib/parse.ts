import { MarkdownLexer } from "./g4/MarkdownLexer";
import { CharStreams, CommonTokenStream } from "antlr4ts";
import {
  BlockCodeContext,
  BlockContext,
  HeaderContext,
  InlineCodeContext,
  InlineHeaderContext,
  MarkdownContext,
  MarkdownParser,
  TextContentContext,
  TextContext,
} from "./g4/MarkdownParser";
import { MarkdownVisitor } from "./g4/MarkdownVisitor";
import { ParseTreeVisitor, ParseTree } from "antlr4";

export function parse(input: string): MarkdownContext {
  const inputStream = CharStreams.fromString(input);
  const lexer = new MarkdownLexer(inputStream);
  const tokens = new CommonTokenStream(lexer);
  const parser = new MarkdownParser(tokens);
  const tree = parser.markdown();

  const visitor = new HelperVisitor();
  const result = visitor.visit(tree as unknown as ParseTree);

  return result;
}

class HelperVisitor
  extends ParseTreeVisitor<any>
  implements MarkdownVisitor<any>
{
  visitMarkdown(ctx: MarkdownContext) {
    // console.log("visitMarkdown");
    return {
      tag: "markdown",
      body: this.visitChildren(ctx),
    };
  }

  visitBlock(ctx: BlockContext) {
    // console.log("visitBlock", ctx);
    return {
      tag: "block",
      body: this.visitChildren(ctx),
    };
  }

  visitInlineHeader(ctx: InlineHeaderContext) {
    // console.log("visitHeader", ctx);
    return {
      tag: "header",
      attr: {
        level: ctx.header().text.length - 1,
      },
      body: this.visitChildren(ctx.textContent()),
    };
  }

  visitTextContent(ctx: TextContentContext) {
    return {
      tag: "text-content",
      body: this.visitChildren(ctx),
    };
  }

  visitText(ctx: TextContext) {
    // console.log("visitText", ctx);
    return {
      tag: "text",
      content: ctx.text,
    };
  }

  visitInlineCode(ctx: InlineCodeContext) {
    // console.log("visitInlineCode", ctx);
    return {
      tag: "inline-code",
      content: ctx.text,
    };
  }

  visitBlockCode(ctx: BlockCodeContext) {
    // console.log("visitBlockCode", ctx);
    const text = ctx.BLOCK_CODE().text;
    const firstLnIndex = text.indexOf("\n");
    const lang = text.substring(3, firstLnIndex);
    const content = text.substring(firstLnIndex + 1, text.length - 3);
    return {
      tag: "block-code",
      attr: {
        lang,
      },
      content,
    };
  }
}
