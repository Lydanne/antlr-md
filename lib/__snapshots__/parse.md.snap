MarkdownNode {
  "attr": {},
  "body": [
    BlockNode {
      "attr": {},
      "body": [
        InlineHeaderNode {
          "attr": {
            "level": 1,
          },
          "body": [
            TextNode {
              "attr": {},
              "body": undefined,
              "content": "你好",
              "tag": "text",
            },
            InlineCodeNode {
              "attr": {},
              "body": undefined,
              "content": "今天",
              "tag": "inline-code",
            },
          ],
          "content": undefined,
          "tag": "inline-header",
        },
      ],
      "content": undefined,
      "tag": "block",
    },
    BlockNode {
      "attr": {},
      "body": [
        InlineTextNode {
          "attr": {},
          "body": [
            TextNode {
              "attr": {},
              "body": undefined,
              "content": "好呀好呀",
              "tag": "text",
            },
          ],
          "content": undefined,
          "tag": "inline-text",
        },
      ],
      "content": undefined,
      "tag": "block",
    },
    BlockNode {
      "attr": {},
      "body": [
        InlineTextNode {
          "attr": {},
          "body": [
            TextNode {
              "attr": {},
              "body": undefined,
              "content": "好呀好呀",
              "tag": "text",
            },
          ],
          "content": undefined,
          "tag": "inline-text",
        },
      ],
      "content": undefined,
      "tag": "block",
    },
  ],
  "content": undefined,
  "tag": "markdown",
}